meshmode/dof_array.py

```diff
@@ -448,34 +463,97 @@ def prg():
         for grp_start, (nel, ndof) in zip(group_starts, group_shapes)))
 
 
-def unflatten(actx: ArrayContext, discr,
-        ary: ArrayContainer,
-        ndofs_per_element_per_group: Optional[Iterable[int]] = None):
-    r"""Convert all 'flat' arrays returned by :func:`flatten` back to
+def _unflatten_group_sizes(discr, ndofs_per_element_per_group):
+    if ndofs_per_element_per_group is None:
+        ndofs_per_element_per_group = [
+                grp.nunit_dofs for grp in discr.groups]
+
+    group_shapes = [
+            (grp.nelements, ndofs_per_element)
+            for grp, ndofs_per_element
+            in zip(discr.groups, ndofs_per_element_per_group)]
+
+    group_sizes = [nel * ndof for nel, ndof in group_shapes]
+    group_starts = np.cumsum([0] + group_sizes)
+
+    return group_shapes, group_starts
+
+
+def unflatten(
+        actx: ArrayContext, discr, ary: ArrayContainer,
+        ndofs_per_element_per_group: Optional[Iterable[int]] = None, *,
+        strict: bool = True,
+        ) -> ArrayContainer:
+    r"""Convert all "flat" arrays returned by :func:`flatten` back to
     :class:`DOFArray`\ s.
 
+    This function recurses into the :class:`~arraycontext.ArrayContainer` for all
+    :class:`DOFArray`\ s. It assumes that all the :class:`DOFArray`\ s inside the
+    container *ary* have the same number of degrees of freedom per
+    element group in *discr*.
```

Comment:
```suggestion
    :class:`DOFArray`\ s. All class:`DOFArray`\ s inside the container 
    *ary* must agree on the mapping from element group number to number
    of degrees of freedom, as given by `ndofs_per_element_per_group`
    (or via *discr*).
```
Thanks for clarifying, I understand now. The original version sounded like all groups needed to have the same number of DOFs per element. I'll commit my version and merge this. Protests if you think I made it worse. :)